packages/react-transform/src/index.ts

```diff
@@ -0,0 +1,350 @@
+		default:
+			return false;
+	}
+}
+
+function shouldTransform(
+	path: NodePath<FunctionLike>,
+	options: PluginOptions
+): boolean {
+	if (getData(path, alreadyTransformed) === true) return false;
+
+	// Opt-out takes first precedence
+	if (isOptedOutOfSignalTracking(path)) return false;
+	// Opt-in opts in to transformation regardless of mode
+	if (isOptedIntoSignalTracking(path)) return true;
+
+	if (options.mode == null || options.mode === "auto") {
+		return (
+			fnNameStartsWithCapital(path) && // Function name indicates it's a component
+			getData(path.scope, maybeUsesSignal) === true && // Function appears to use signals
+			getData(path.scope, containsJSX) === true && // Function contains JSX
+			path.scope.parent === path.scope.getProgramParent() // Function is top-level
+		);
+	}
+
+	return false;
+}
+
+function isValueMemberExpression(
+	path: NodePath<BabelTypes.MemberExpression>
+): boolean {
+	return (
+		(path.node.property.type === "Identifier" &&
+			path.node.property.name === "value") ||
+		(path.node.property.type === "StringLiteral" &&
+			path.node.property.value === "value")
+	);
+}
+
+const tryCatchTemplate = template.statements`var STOP_TRACKING_IDENTIFIER = HOOK_IDENTIFIER();
+try {
+	BODY
+} finally {
+	STOP_TRACKING_IDENTIFIER();
+}`;
+
+export interface PluginOptions {
+	/**
+	 * Specify the mode to use:
+	 * - `auto`: Automatically wrap all components that use signals.
+	 * - `manual`: Only wrap components that are annotated with `@trackSignals` in a JSX comment.
+	 */
+	mode?: "auto" | "manual";
+	experimental?: {
+		/**
+		 * If set to true, the component body will not be wrapped in a try/finally
+		 * block and instead the next component render or a microtick will stop
+		 * tracking signals for this component. This is an experimental feature and
+		 * may be removed in the future.
+		 * @default false
+		 */
+		noTryFinally?: boolean;
+	};
+}
+
+export default function signalsTransform(
+	{ types: t }: PluginArgs,
+	options: PluginOptions
+): PluginObj {
+	return {
+		name: "@preact/signals-transform",
+		visitor: {
+			Program: {
+				enter(path, state) {
+					// Following the pattern of babel-plugin-transform-react-jsx, we
+					// lazily create the import statement for the useSignalTracking hook.
+					// We create a function and store it in the PluginPass object, so that
+					// on the first usage of the hook, we can create the import statement.
+					set(
+						state,
+						getHookIdentifier,
+						createImportLazily(t, state, path, importName, importSource)
+					);
+				},
+			},
+
+			ArrowFunctionExpression: {
+				// TODO: Consider alternate implementation, where on enter of a function
+				// expression, we run our own manual scan the AST to determine if the
+				// function uses signals and is a component. This manual scan once upon
+				// seeing a function would probably be faster than running an entire
+				// babel pass with plugins on components twice.
```

Comment:
For anyone with more experience with babel transforms - if I want to know if a function body contains JSX to know whether to perform my transformation, is it better to do a local visitor pass inside this ArrowFunctionExpression handler to observe that (only running my local visitor), and then transform? Or to do the observation in the larger babel visitor pass and then transform after, which triggers and additional visitor pass with all plugins?

Reply:
That's a good question. I've never thought about this.

Reply:
You can use something like [this](https://github.com/preactjs/prefresh/blob/main/packages/babel/src/index.mjs#L168-L183) on a VariableDeclaration to see whether they are used as JSX

Reply:
Oh interesting. IIUC, this code looks to see if a particular identifier is used as JSX somewhere in the same file?

Reply:
ScopeBindings should work across files afaik